SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -375,32 +436,164 @@ - (void)updateProfileOnService:(nullable NSString *)localProfileName
+{
+    OWSAssert(profileKey.length == kProfileKeyLength);
+    OWSAssert(recipientId.length > 0);
+    if (profileKey.length != kProfileKeyLength) {
+        return;
+    }
+
+    NSData *_Nullable existingProfileKey = [self profileKeyForRecipientId:recipientId];
+    if (existingProfileKey &&
+        [existingProfileKey isEqual:profileKey]) {
+        // Ignore redundant update.
+        return;
+    }
+
+    [self.dbConnection setObject:profileKey
+                          forKey:recipientId
+                    inCollection:kOWSProfilesManager_OtherUsersProfileKeysCollection];
+    self.otherUsersProfileKeyCache[recipientId] = profileKey;
+}
+
+- (nullable NSData *)profileKeyForRecipientId:(NSString *)recipientId
+{
+    OWSAssert(recipientId.length > 0);
+
+    NSData *_Nullable profileKey = self.otherUsersProfileKeyCache[recipientId];
+    if (profileKey.length > 0) {
+        return profileKey;
+    }
+
+    profileKey =
+        [self.dbConnection objectForKey:recipientId inCollection:kOWSProfilesManager_OtherUsersProfileKeysCollection];
+    if (profileKey) {
+        OWSAssert(profileKey.length == kProfileKeyLength);
+        self.otherUsersProfileKeyCache[recipientId] = profileKey;
+    }
+    return profileKey;
+}
+
+- (nullable NSString *)profileNameForRecipientId:(NSString *)recipientId
+{
+    OWSAssert([NSThread isMainThread]);
+    OWSAssert(recipientId.length > 0);
+
+    [self fetchProfileForRecipientId:recipientId];
+
+    UserProfile *userProfile = [self getOrCreateUserProfileForRecipientId:recipientId];
+    return userProfile.profileName;
+}
+
+- (nullable UIImage *)profileAvatarForRecipientId:(NSString *)recipientId
+{
+    OWSAssert([NSThread isMainThread]);
+    OWSAssert(recipientId.length > 0);
+
+    [self fetchProfileForRecipientId:recipientId];
+
+    UIImage *_Nullable image = [self.otherUsersProfileAvatarImageCache objectForKey:recipientId];
+    if (image) {
+        return image;
+    }
+
+    UserProfile *userProfile = [self getOrCreateUserProfileForRecipientId:recipientId];
+    if (userProfile.avatarFileName) {
+        image = [self loadProfileAvatarWithFilename:userProfile.avatarFileName];
+        if (image) {
+            [self.otherUsersProfileAvatarImageCache setObject:image forKey:recipientId];
         }
+    }
 
-        dispatch_async(dispatch_get_main_queue(), ^{
-            self.localProfileName = localProfileName;
-            self.localProfileAvatarImage = localProfileAvatarImage;
-            self.localProfileAvatarMetadata = localProfileAvatarMetadata;
+    return image;
+}
 
-            [[NSNotificationCenter defaultCenter] postNotificationName:kNSNotificationName_LocalProfileDidChange
-                                                                object:nil
-                                                              userInfo:nil];
-        });
-    });
+- (void)fetchProfileForRecipientId:(NSString *)recipientId
```

Comment:
I'm not sure why we need this method. 

My thought was the ProfileFetcherJob is run whenever a thread becomes visible (It's already throttled).
Whenever it fetches it tells the ProfileManager the latest details for a profile.
ProfileManager can expire cache immediately if the profile has changed.